resharper/resharper-unity/src/Rider/UnitTesting/RunViaUnityEditorStrategy.cs

```diff
@@ -428,31 +441,38 @@ private void SubscribeResults(IUnitTestRun firstRun, Lifetime connectionLifetime
 
         private Task WaitForUnityEditorConnectedAndIdle(Lifetime lifetime, CancellationToken cancellationToken)
         {
+            myLogger.Trace("WaitForUnityEditorConnectedAndIdle");
             var tcs = new TaskCompletionSource<Unit>();
-
             var waitingLifetimeDef = Lifetime.Define(lifetime);
-            waitingLifetimeDef.Lifetime.OnTermination(() => tcs.TrySetCanceled());
-
+            var waitingLifetime = waitingLifetimeDef.Lifetime;
+            waitingLifetime.OnTermination(() => tcs.TrySetCanceled());
             tcs.Task.ContinueWith(_ => waitingLifetimeDef.Terminate(), lifetime);
+            if (cancellationToken.IsCancellationRequested) tcs.TrySetCanceled();
 
-            mySolution.Locks.ExecuteOrQueue(lifetime, "WaitForUnityEditorConnectedAndIdle", () =>
+            // Can't use ConnectionTracker because it pauses without focus
+            waitingLifetime.StartMainUnguardedAsync(async () =>
```

Comment:
@kskrygan, I am not sure about this code (while loop inside async main activity), could you take a look?

Reply:
@van800 @krasnotsvetov I would suggest using 
JetBrains.Application.Threading.Tasks.TaskHostEx.QueueRecuring
instead of a while loop. Please rework this part

Reply:
QueueRecuring only fires, when Rider has focus. While after running tests, some users what to do something else.